File: app/ReleaseImageGenerator.Domain/ColorGenerator.cs
using Wacton.Unicolour;

namespace ReleaseImageGenerator.Domain;

public static class ColorGenerator
{
    public enum ColorLimitation
    {
        ALL,
        LIGHTER,
        NEUTRAL_LIGHTNESS,
        DARKER,
        SATURATED,
        NEUTRAL_SATURATION,
        DESATURATED,
    }

    public static Unicolour GetRandomColor(params ColorLimitation[] limitations)
    {
        var random = new Random();
        var saturationRange = limitations?.Aggregate((0D, 1D), (current, limitation) => limitation switch
        {
            ColorLimitation.NEUTRAL_SATURATION => (0.25, 0.75),
            ColorLimitation.SATURATED => (0.5, 1D),
            ColorLimitation.DESATURATED => (0D, 0.5),
            _ => current
        }) ?? (0D, 1D);
        var lightnessRange = limitations?.Aggregate((0D, 1D), (current, limitation) => limitation switch
        {
            ColorLimitation.NEUTRAL_LIGHTNESS => (0.25, 0.75),
            ColorLimitation.LIGHTER => (0.5, 1D),
            ColorLimitation.DARKER => (0D, 0.5),
            _ => current
        }) ?? (0D, 1D);
        var color = new Unicolour(ColourSpace.Oklch,
            (lightnessRange.Item1 + random.NextDouble() * (lightnessRange.Item2 - lightnessRange.Item1 + double.Epsilon),
                saturationRange.Item1 +
                random.NextDouble() * (saturationRange.Item2 - saturationRange.Item1 + double.Epsilon),
                random.Next(0, 255)
            )).MapToRgbGamut();
        return color;
    }

    public static Unicolour[] GetRandomPalette(Unicolour primaryColor, int numberOfColors = 5, int spread = 20, double alpha = 1D)
    {
        // Ensure we get an odd number of colors for better balance
        var halfPaletteSize = numberOfColors / 2;
        var colors = new List<Unicolour>();

        // Get the HSL representation of the primary color
        var primaryHsl = primaryColor.Hsl;

        for (int i = -halfPaletteSize; i <= halfPaletteSize; i++)
        {
            // Create a new color with the shifted hue
            var shiftedColor = new Unicolour(ColourSpace.Hsl, primaryHsl.H + i * spread, primaryHsl.S, primaryHsl.L, alpha);
            colors.Add(shiftedColor);
        }

        return colors.ToArray();
    }

    public static Unicolour[] GetRandomPalette(int numberOfColors) =>
        GetRandomPalette(GetRandomColor(), numberOfColors);

    public static Unicolour[] GetRandomPalette() => GetRandomPalette(GetRandomColor());
}

File: src/lib/ColorGenerator.ts
import type { Color, OklchColor, RgbColor } from "./types";
import { parse, oklch, converter } from "culori";

export enum ColorLimitation {
  ALL = "ALL",
  LIGHTER = "LIGHTER",
  NEUTRAL_LIGHTNESS = "NEUTRAL_LIGHTNESS",
  DARKER = "DARKER",
  SATURATED = "SATURATED",
  NEUTRAL_SATURATION = "NEUTRAL_SATURATION",
  DESATURATED = "DESATURATED",
}

export class ColorGenerator {
  static getRandomColor(...limitations: ColorLimitation[]): Color {
    let saturationRange = { min: 0, max: 0.4 };
    let lightnessRange = { min: 0, max: 1 };

    limitations.forEach((limitation) => {
      switch (limitation) {
        case ColorLimitation.NEUTRAL_SATURATION:
          saturationRange = { min: 0.1, max: 0.3 };
          break;
        case ColorLimitation.SATURATED:
          saturationRange = { min: 0.2, max: 0.4 };
          break;
        case ColorLimitation.DESATURATED:
          saturationRange = { min: 0, max: 0.2 };
          break;
        case ColorLimitation.NEUTRAL_LIGHTNESS:
          lightnessRange = { min: 0.25, max: 0.75 };
          break;
        case ColorLimitation.LIGHTER:
          lightnessRange = { min: 0.5, max: 1 };
          break;
        case ColorLimitation.DARKER:
          lightnessRange = { min: 0, max: 0.5 };
          break;
      }
    });

    const l =
      lightnessRange.min +
      Math.random() * (lightnessRange.max - lightnessRange.min);
    const c =
      saturationRange.min +
      Math.random() * (saturationRange.max - saturationRange.min);
    const h = Math.random() * 360;

    return this.createColor(l, c, h, 1);
  }

  static createColor(
    l: number,
    c: number,
    h: number,
    alpha: number = 1,
  ): Color {
    const oklch = { l, c, h };
    const rgb = this.oklchToRgb(oklch);

    return { oklch, rgb, alpha };
  }

  static parseColor(colorString: string): Color {
    const parsed = parse(colorString);
    if (!parsed) {
      return this.getRandomColor(
        ColorLimitation.NEUTRAL_LIGHTNESS,
        ColorLimitation.NEUTRAL_SATURATION,
      );
    }

    const oklchClr = oklch(parsed);
    return this.createColor(
      oklchClr.l || 0.5,
      oklchClr.c || 0.1,
      oklchClr.h || 0,
      oklchClr.alpha ?? 1,
    );
  }

  static getRandomPalette(
    primaryColor: Color,
    numberOfColors: number = 5,
    spread: number = 20,
    alpha: number = 1,
  ): Color[] {
    const halfPaletteSize = Math.floor(numberOfColors / 2);
    const colors: Color[] = [];

    for (let i = -halfPaletteSize; i <= halfPaletteSize; i++) {
      const shiftedHue = (primaryColor.oklch.h + i * spread + 360) % 360;
      colors.push(
        this.createColor(
          primaryColor.oklch.l,
          primaryColor.oklch.c,
          shiftedHue,
          alpha,
        ),
      );
    }

    return colors;
  }

  static oklchToRgb(oklchClr: OklchColor): RgbColor {
    const toRgb = converter("rgb");
    const oklchString = `oklch(${oklchClr.l} ${oklchClr.c} ${oklchClr.h})`;
    const rgbClr = toRgb(oklchString);

    return {
      r: Math.max(0, Math.min(1, rgbClr?.r ?? 0)),
      g: Math.max(0, Math.min(1, rgbClr?.g ?? 0)),
      b: Math.max(0, Math.min(1, rgbClr?.b ?? 0)),
    };
  }

  static colorToRgbaString(color: Color): string {
    return `rgba(${Math.round(color.rgb.r * 255)}, ${Math.round(color.rgb.g * 255)}, ${Math.round(color.rgb.b * 255)}, ${color.alpha})`;
  }

  static colorToRgbString(color: Color): string {
    return `rgb(${Math.round(color.rgb.r * 255)}, ${Math.round(color.rgb.g * 255)}, ${Math.round(color.rgb.b * 255)})`;
  }
}
